Objective-C/CBLDatabase.mm

```diff
@@ -864,139 +870,201 @@ - (CBLDocument*) saveDocument: (CBLDocument*)document
+                doc = deletedDoc;
+            }
+            
+            if (doc.c4Doc)
+                baseDoc = [[CBLDocument alloc] initWithDatabase: self
+                                                     documentID: docID
+                                                          c4Doc: doc.c4Doc];
+            
+            otherDoc = [[CBLDocument alloc] initWithDatabase: self
+                                                  documentID: docID
+                                              includeDeleted: YES
+                                                       error: outError];
+            if (!otherDoc)
+                return nil;
+            
+            // Abort transaction:
+            if (!transaction.abort()) {
+                convertError(transaction.error(), outError);
+                return nil;
+            }
+        }
         
-        // Now save the merged properties:
-        if (![self saveDocument: resolved
-                           into: &newDoc
-                       withBase: current.c4Doc.rawDoc
-                     asDeletion: deletion
-                          error: outError])
+        // Resolve conflict:
+        id<CBLConflictResolver> resolver = self.config.conflictResolver;
+        CBLConflict* conflict = [[CBLConflict alloc] initWithMine: doc
+                                                           theirs: otherDoc
+                                                             base: baseDoc];
+        CBLDocument* resolved = [resolver resolve: conflict];
+        if (!resolved) {
+            convertError({LiteCoreDomain, kC4ErrorConflict}, outError);
             return nil;
-        Assert(newDoc); // In a transaction we can't have a second conflict after merging!
-    }
-    
-    // Save succeeded; now commit the transaction:
-    if (!transaction.commit()) {
-        c4doc_free(newDoc);
-        convertError(transaction.error(), outError);
-        return nil;
+        }
+        
+        CBL_LOCK(self) {
+            CBLDocument* current = [[CBLDocument alloc] initWithDatabase: self
+                                                              documentID: docID
+                                                          includeDeleted: YES
+                                                                   error: outError];
+            if ($equal(resolved.revID, current.revID))
+                return resolved; // Same as current
+            
+            // For saving:
+            doc = resolved;
+            baseRev = current.c4Doc.rawDoc;
+            deletion = resolved.isDeleted;
+        }
     }
-    
-    CBLC4Document* c4Doc = [CBLC4Document document: newDoc];
-    return [[CBLDocument alloc] initWithDatabase: self
-                                      documentID: document.id
-                                           c4Doc: c4Doc];
 }
 
 
 #pragma mark - RESOLVING REPLICATED CONFLICTS:
 
 
+// Called inside the replicator's database transaction.
```

Comment:
I don't think this is true; the calling method doesn't open a transaction, and it's called from dispatch_async.